test/prefer-dataset.js

```diff
@@ -8,18 +9,67 @@ const ruleTester = avaRuleTester(test, {
 	}
 });
 
+const errors = [
+	{
+		ruleId: 'prefer-dataset',
+		message: 'Prefer `dataset` over `setAttribute`'
+	}
+];
+
 ruleTester.run('prefer-dataset', rule, {
 	valid: [
 		'element.dataset.unicorn = \'🦄\';',
 		'element.dataset[\'unicorn\'] = \'🦄\';',
 		'element[setAttribute](\'data-unicorn\', \'🦄\');',
-		'element.setAttribute(\'foo\', \'bar\');'
+		'element.setAttribute(\'foo\', \'bar\');',
+		'element.setAttribute(foo, bar);',
+		'element.getAttribute(\'data-unicorn\');'
 	],
 	invalid: [
 		{
 			code: 'element.setAttribute(\'data-unicorn\', \'🦄\');',
-			errors: [{message: 'Prefer `dataset` over `setAttribute`'}],
+			errors,
 			output: 'element.dataset.unicorn = \'🦄\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-🦄\', \'🦄\');',
+			errors,
+			output: 'element.dataset.🦄 = \'🦄\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-foo2\', \'🦄\');',
+			errors,
+			output: 'element.dataset.foo2 = \'🦄\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-foo:bar\', \'zaz\');',
+			errors,
+			output: 'element.dataset[\'foo:bar\'] = \'zaz\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-foo.bar\', \'zaz\');',
+			errors,
+			output: 'element.dataset[\'foo.bar\'] = \'zaz\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-foo-bar\', \'zaz\');',
+			errors,
+			output: 'element.dataset.fooBar = \'zaz\';'
+		},
+		{
+			code: 'element.setAttribute(\'data-foo\', /* comment */ \'bar\');',
+			errors,
+			output: 'element.dataset.foo = \'bar\';'
```

Comment:
I don't think we want to delete comments during fixing but it probably doesn't come up that often. @sindresorhus, what's your take on this?

1. Removing comments is okay
2. Comments should be preserved (may be non-trivial)
3. Do not fix if comments are detected

[I've done (3) in the past](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/305/files#diff-c717db2a17e1fd6455af92ec118c404eR35) but I think a bunch of different rules current do (1).

Reply:
@MrHen I don't think we should do `1.`, but we need to come up with a guideline that we can apply to all rules. Maybe we could make an utility that just extract all comments to above a node we give it? The comments can require manual formatting and movement. We just shouldn't remove them. Alternatively, `3.` if `2.` is too complex, but I think we could make something generic that we could apply to many rules. Would you be able to open a new issue about it? We probably need to review all existing rules for this and update the contributing guidelines.

Reply:
https://github.com/sindresorhus/eslint-plugin-unicorn/issues/370